src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs

```diff
@@ -0,0 +1,264 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Roslyn.Utilities;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.Diagnostics
+{
+    /// <summary>
+    /// Root type for both document and workspace diagnostic pull requests.
+    /// </summary>
+    internal abstract class AbstractPullDiagnosticHandler<TDiagnosticsParams, TReport> : IRequestHandler<TDiagnosticsParams, TReport[]?>
```

Comment:
This handler will not be called from the host because the LSP server in which it lives is disabled locally
http://sourceroslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/LanguageClient/LiveShareLanguageServerClient.cs,21

Ignoring razor, the only LSP server that runs locally is the OOP LSP server here - http://sourceroslyn.io/#Microsoft.CodeAnalysis.Remote.ServiceHub/Services/LanguageServer/LanguageServer.cs,50

This server can be updated with the correct initialize response and needs to have an LSP entry point for pull diagnostics similar to http://sourceroslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/LanguageClient/InProcLanguageServer.cs,a3c05b85bc951abd 
In general this LSP server should be updated to use the LSP handler infrastructure (since this is the first feature we're moving it hasn't been updated).